syntax = "proto3";

package paxos;

import "google/protobuf/empty.proto";

service Acceptor {
    rpc Prepare(PrepareRequest) returns (PrepareResponse) {}
    rpc Accept(AcceptRequest) returns (AcceptResponse) {}
    rpc SendPing(PingRequest) returns (PingResponse) {}
}

message PrepareRequest {
    uint64 proposal_number = 1;
}

message PrepareResponse {
    bool has_accepted_value = 1;
    uint64 accepted_proposal = 2;
    string accepted_value = 3;
    uint64 min_proposal = 4;
}

message AcceptRequest {
    uint64 proposal_number = 1;
    string value = 2;
}

message AcceptResponse {
    uint64 min_proposal = 1;
}

message PingRequest {
    uint32 node_id = 1;
}

message PingResponse {
    uint32 node_id = 1;
}